### ISSUE TITLE
fragpipe in linux when building library using easypqp

### ISSUE THREAD

**Issue by zhutiansheng** [Nesvilab/FragPipe#210]
**Describe the bug**
when using window fragpipe, the command in log file is:

d:\Users\sky\anaconda3\python -u C:\GreenProgram\fragpipe13\lib\..\tools\speclib\gen_con_spec_lib.py D:\FragPipe\fas\2020-07-17-decoys-UP000005640_reviewed_human_20200401_irt_fragpipe.fasta D:\temp D:\FragPipe\data\X20181025yix_HCC_SCX_DDA_7A.mzXML;D:\FragPipe\data\X20181025yix_HCC_SCX_DDA_6B.mzXML;D:\FragPipe\data\X20181025yix_HCC_SCX_DDA_7B.mzXML D:\temp True unused use_easypqp ciRT 7 "--rt_lowess_frac
[log_2020-07-20_13-37-58.txt](https://github.com/Nesvilab/FragPipe/files/4945976/log_2020-07-20_13-37-58.txt)
tion 0.01"

and it runs well. But when in linux (the same version of fragpipe and msfrage and phli* ), the command is :

python -u ../fragpipe13/tools/speclib/gen_con_spec_lib.py ../FragPip/test/fas/2020-05-30-decoys-UP000005640_reviewed_human_20200401_irt_rev.fasta ./ ../FragPip/test/mzXML//X20181025yix_HCC_SCX_DDA_6B.mzXML;../FragPip/test/mzXML//X20181025yix_HCC_SCX_DDA_7A.mzXML;../FragPip/test/mzXML//X20181025yix_HCC_SCX_DDA_7B.mzXML output True unused use_easypqp ciRT 7 "--rt_lowess_fraction 0.01"

it return the error: 
 File "../fragpipe13/tools/speclib/gen_con_spec_lib.py", line 5
    use_philosopher_fo: bool = not True
                      ^
SyntaxError: invalid syntax
../FragPip/test/mzXML//X20181025yix_HCC_SCX_DDA_7A.mzXML: line 1: syntax error near unexpected token `newline'
../FragPip/test/mzXML//X20181025yix_HCC_SCX_DDA_7A.mzXML: line 1: `<?xml version="1.0" encoding="ISO-8859-1"?>'
../FragPip/test/mzXML//X20181025yix_HCC_SCX_DDA_7B.mzXML: line 1: syntax error near unexpected token `newline'
../FragPip/test/mzXML//X20181025yix_HCC_SCX_DDA_7B.mzXML: line 1: `<?xml version="1.0" encoding="ISO-8859-1"?>'

why?

**Comment by guoci** [comment]
You need to do the following on Linux:
1. use Anaconda Python 3 (other Python version may work, but Anocanda is well tested.)
2. for the command, it is specific to Windows, so on Linux replace it with:
```
python -u ../fragpipe13/tools/speclib/gen_con_spec_lib.py ../FragPip/test/fas/2020-05-30-decoys-UP000005640_reviewed_human_20200401_irt_rev.fasta ./ ../FragPip/test/mzXML/X20181025yix_HCC_SCX_DDA_6B.mzXML:../FragPip/test/mzXML/X20181025yix_HCC_SCX_DDA_7A.mzXML:../FragPip/test/mzXML/X20181025yix_HCC_SCX_DDA_7B.mzXML output True unused use_easypqp ciRT 7 --rt_lowess_fraction 0.01
```

**Comment by zhutiansheng** [comment]
I used the command: docker pull grosenberger/easypqp:latest
And ran the docker and installed lxml package in the docker container。
Then I ran the command you provided, got the following errors: 
Traceback (most recent call last):
  File "../fragpipe13/tools/speclib/gen_con_spec_lib.py", line 155, in <module>
    easypqp = get_bin_path('easypqp', 'easypqp')
  File "../fragpipe13/tools/speclib/gen_con_spec_lib.py", line 68, in get_bin_path
    rel_loc, = [e for e in dist.get('files') if pathlib.Path(e).stem == bin_stem]
TypeError: 'NoneType' object is not iterable

root@2c348bc0f79a:/mnt/data/fragpipeWorkspace# which easypqp
/usr/local/bin/easypqp

**Comment by guoci** [comment]
@zhutiansheng that doesn't work, you need to install EasyPQP with `pip`

**Comment by zhutiansheng** [comment]
thanks

**Comment by plbaldoni** [comment]
> You need to do the following on Linux:
> 
> 1. use Anaconda Python 3 (other Python version may work, but Anocanda is well tested.)
> 2. for the command, it is specific to Windows, so on Linux replace it with:
> 
> ```
> python -u ../fragpipe13/tools/speclib/gen_con_spec_lib.py ../FragPip/test/fas/2020-05-30-decoys-UP000005640_reviewed_human_20200401_irt_rev.fasta ./ ../FragPip/test/mzXML/X20181025yix_HCC_SCX_DDA_6B.mzXML:../FragPip/test/mzXML/X20181025yix_HCC_SCX_DDA_7A.mzXML:../FragPip/test/mzXML/X20181025yix_HCC_SCX_DDA_7B.mzXML output True unused use_easypqp ciRT 7 --rt_lowess_fraction 0.01
> ```

Apologies for revisiting this closed issue, but I am experiencing a similar issue with the new version of FragPipe.

What is the correct syntax to call easypqp to generate spectral libraries on Linux after running Philosopher `filter`? Translating from Windows to Linux, it would look like this:

```
$pythonPath -u $specLibPath $fastaPath $PWD unused $PWD True unused use_easypqp "ciRT:noIM" 13 "--max_delta_unimod 0.02 --max_delta_ppm 15.0 --fragment_types [\'b\',\'y\',]" "--rt_lowess_fraction 0.0" delete_intermediate_files $PWD/*_uncalibrated.mgf
```

Could someone help me what is wrong with the above? All uncalibrated mgfs are located in $PWD. From the error message, it seems that `gen_con_spec_lib.py` is trying to find `filter.log`, which does not exist.

**Comment by fcyu** [comment]
Guo Ci @guoci , can you take a look and answer the questions when you have time?

Thanks,

Fengchao

**Comment by guoci** [comment]
Hi @plbaldoni,
Please post the log file and also use the latest version of FragPipe.

**Comment by plbaldoni** [comment]
Thanks @guoci , the log and the .sh file with the sequence of commands is [here](https://github.com/Nesvilab/FragPipe/files/7073732/fragpipe.zip)

**Comment by guoci** [comment]
Hi @plbaldoni,
In the `philosopher filter` command, you need to write the output to `filter.log`.

**Comment by plbaldoni** [comment]
Thanks @guoci . After redirecting the output with `philosopher filter ... |& tee filter.log` (using the same syntax as in my previous [post](https://github.com/Nesvilab/FragPipe/issues/210#issuecomment-907957198) ), I get the following:

```
(base) [baldoni.p@milton-sml-023 tmp]$ $pythonPath -u $specLibPath $fastaPath $PWD/ $pepFiles $PWD/ True $philosopherPath use_easypqp "noirt:noim" $num_threads "--max_delta_unimod 0.02 --max_delta_ppm 15.0 --fragment_types [\'b\',\'y\',]" "--rt_lowess_fraction 0.0"

Spectral library building
Commands to execute:
/home/anaconda3/envs/py37/bin/easypqp library --psmtsv /project/fragpipe/PXD019086/Raw_Celegans/tmp/psm.tsv --peptidetsv /project/fragpipe/PXD019086/Raw_Celegans/tmp/peptide.tsv --out easypqp_lib_openswath.tsv --rt_lowess_fraction 0.0 filelist_easypqp_library.txt
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
easypqp, version 0.1.25
Error: No PSMs files present. Need to have tag 'psmpkl' in filename.
Library not generated, not enough peptides could be found for alignment.
Please try using other options for alignment (e.g. ciRT if used other options)
```

The file `filelist_easypqp_library.txt` is empty, and it looks like `easypqp convert` is not being triggered. Am I misspecifying the call of `gen_con_spec_lib.py`?

This is what `filter.log` looks like:

```
(base) [baldoni.p@milton-sml-023 tmp]$ cat filter.log 
time="10:53:45" level=info msg="Executing Filter  v4.0.0"
time="10:53:45" level=info msg="Processing peptide identification files"
time="10:53:49" level=info msg="1+ Charge profile" decoy=7 target=70
time="10:53:49" level=info msg="2+ Charge profile" decoy=997 target=22147
time="10:53:49" level=info msg="3+ Charge profile" decoy=100 target=3980
time="10:53:49" level=info msg="4+ Charge profile" decoy=12 target=253
time="10:53:49" level=info msg="5+ Charge profile" decoy=0 target=0
time="10:53:49" level=info msg="6+ Charge profile" decoy=0 target=0
time="10:53:49" level=info msg="Database search results" ions=7972 peptides=7076 psms=27566
time="10:53:50" level=info msg="Converged to 1.00 % FDR with 23774 PSMs" decoy=239 threshold=0.5431 total=24013
time="10:53:51" level=info msg="Converged to 0.99 % FDR with 4607 Peptides" decoy=46 threshold=0.9414 total=4653
time="10:53:51" level=info msg="Converged to 0.99 % FDR with 5410 Ions" decoy=54 threshold=0.9345 total=5464
time="10:53:52" level=info msg="Protein inference results" decoy=849 target=2139
time="10:53:53" level=info msg="Converged to 1.03 % FDR with 1645 Proteins" decoy=17 threshold=0.981 total=1662
time="10:53:55" level=info msg="Applying sequential FDR estimation" ions=5791 peptides=4948 psms=23574
time="10:53:55" level=info msg="Converged to 0.11 % FDR with 23547 PSMs" decoy=27 threshold=0.5431 total=23574
time="10:53:56" level=info msg="Converged to 0.34 % FDR with 4931 Peptides" decoy=17 threshold=0.5481 total=4948
time="10:53:56" level=info msg="Converged to 0.29 % FDR with 5774 Ions" decoy=17 threshold=0.5481 total=5791
time="10:53:57" level=info msg="Post processing identifications"
time="10:53:59" level=info msg="Assigning protein identifications to layers"
time="10:53:59" level=info msg="Processing protein inference"
time="10:54:01" level=info msg="Synchronizing PSMs and proteins"
time="10:54:01" level=info msg=Saving
time="10:54:05" level=info msg=Done
```

**Comment by guoci** [comment]
The last command is incorrect, there is a missing last argument `filelist_speclibgen.txt` which is the filename of a file list of LCMS files. I would recommend you to use FragPipe directly instead of copying the command lines from it.
There are operations done by FragPipe that is not in the command lines. For example, FragPipe generates file list(s).

**Comment by anesvi** [comment]
Guo Ci, maybe you can post the format of the list they are missing. If they already went through the trouble of converting commands to a script. But yes, best to run via FragPipe, which one can run on linux
Alexey

**Comment by guoci** [comment]
@plbaldoni To generate `filelist_speclibgen.txt`, put all the LCMS file paths in it, separated by newlines. Example:
```
/path/to/1.mgf
/path/to/2.mgf
/path/to/3.mgf
```
Can I ask why do you prefer using the command line? Is it due to X11 forwarding being disabled?

**Comment by plbaldoni** [comment]
Thanks @guoci , I appreciate it. The X11 forwarding is definitely an issue for me given the HPC settings that I currently work on. But my preference was mostly due to (1) better automation using the command line and (2) access of better computing resources than the Windows virtual machines that I currently have access to.

**Comment by lzw150** [comment]
> Thanks @guoci . After redirecting the output with `philosopher filter ... |& tee filter.log` (using the same syntax as in my previous [post](https://github.com/Nesvilab/FragPipe/issues/210#issuecomment-907957198) ), I get the following:
> 
> ```
> (base) [baldoni.p@milton-sml-023 tmp]$ $pythonPath -u $specLibPath $fastaPath $PWD/ $pepFiles $PWD/ True $philosopherPath use_easypqp "noirt:noim" $num_threads "--max_delta_unimod 0.02 --max_delta_ppm 15.0 --fragment_types [\'b\',\'y\',]" "--rt_lowess_fraction 0.0"
> 
> Spectral library building
> Commands to execute:
> /home/anaconda3/envs/py37/bin/easypqp library --psmtsv /project/fragpipe/PXD019086/Raw_Celegans/tmp/psm.tsv --peptidetsv /project/fragpipe/PXD019086/Raw_Celegans/tmp/peptide.tsv --out easypqp_lib_openswath.tsv --rt_lowess_fraction 0.0 filelist_easypqp_library.txt
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> easypqp, version 0.1.25
> Error: No PSMs files present. Need to have tag 'psmpkl' in filename.
> Library not generated, not enough peptides could be found for alignment.
> Please try using other options for alignment (e.g. ciRT if used other options)
> ```
> 
> The file `filelist_easypqp_library.txt` is empty, and it looks like `easypqp convert` is not being triggered. Am I misspecifying the call of `gen_con_spec_lib.py`?
> 
> This is what `filter.log` looks like:
> 
> ```
> (base) [baldoni.p@milton-sml-023 tmp]$ cat filter.log 
> time="10:53:45" level=info msg="Executing Filter  v4.0.0"
> time="10:53:45" level=info msg="Processing peptide identification files"
> time="10:53:49" level=info msg="1+ Charge profile" decoy=7 target=70
> time="10:53:49" level=info msg="2+ Charge profile" decoy=997 target=22147
> time="10:53:49" level=info msg="3+ Charge profile" decoy=100 target=3980
> time="10:53:49" level=info msg="4+ Charge profile" decoy=12 target=253
> time="10:53:49" level=info msg="5+ Charge profile" decoy=0 target=0
> time="10:53:49" level=info msg="6+ Charge profile" decoy=0 target=0
> time="10:53:49" level=info msg="Database search results" ions=7972 peptides=7076 psms=27566
> time="10:53:50" level=info msg="Converged to 1.00 % FDR with 23774 PSMs" decoy=239 threshold=0.5431 total=24013
> time="10:53:51" level=info msg="Converged to 0.99 % FDR with 4607 Peptides" decoy=46 threshold=0.9414 total=4653
> time="10:53:51" level=info msg="Converged to 0.99 % FDR with 5410 Ions" decoy=54 threshold=0.9345 total=5464
> time="10:53:52" level=info msg="Protein inference results" decoy=849 target=2139
> time="10:53:53" level=info msg="Converged to 1.03 % FDR with 1645 Proteins" decoy=17 threshold=0.981 total=1662
> time="10:53:55" level=info msg="Applying sequential FDR estimation" ions=5791 peptides=4948 psms=23574
> time="10:53:55" level=info msg="Converged to 0.11 % FDR with 23547 PSMs" decoy=27 threshold=0.5431 total=23574
> time="10:53:56" level=info msg="Converged to 0.34 % FDR with 4931 Peptides" decoy=17 threshold=0.5481 total=4948
> time="10:53:56" level=info msg="Converged to 0.29 % FDR with 5774 Ions" decoy=17 threshold=0.5481 total=5791
> time="10:53:57" level=info msg="Post processing identifications"
> time="10:53:59" level=info msg="Assigning protein identifications to layers"
> time="10:53:59" level=info msg="Processing protein inference"
> time="10:54:01" level=info msg="Synchronizing PSMs and proteins"
> time="10:54:01" level=info msg=Saving
> time="10:54:05" level=info msg=Done
> ```

Have you solved this problem? I have the same problem

**Comment by fcyu** [comment]
@lzw150 , this post is quite old. Are you using the latest versions of the tools? If yes, could you upload your log file to this issue page: https://github.com/Nesvilab/FragPipe/issues/210 

Thanks,

Fengchao

**Comment by lzw150** [comment]
> @lzw150 , this post is quite old. Are you using the latest versions of the tools? If yes, could you upload your log file to this issue page: #210
> 
> Thanks,
> 
> Fengchao

[slurm-54511223.log](https://github.com/Nesvilab/FragPipe/files/13311125/slurm-54511223.log)

I use v19.1, a recent version.

**Comment by fcyu** [comment]
Are you using your own shell script to run the tools? I see that there are warnings from MSFragger saying that some parameters were missing. 

And you are using PeptideProphet to process all pepXML files combined, which is not recommended for closes search.

And you did not run the EasyPQP convert command, which directly caused your error.

We don't recommend running the individual tools using custom shell script because there are changes/adjustments/optimizations under-the-hood in FragPipe that most users won't know.

Please download the latest FragPipe from https://github.com/Nesvilab/FragPipe/releases/latest, and run it in GUI or command line mode.

Best,

Fengchao

**Comment by lzw150** [comment]
> Are you using your own shell script to run the tools? I see that there are warnings from MSFragger saying that some parameters were missing.
> 
> And you are using PeptideProphet to process all pepXML files combined, which is not recommended for closes search.
> 
> And you did not run the EasyPQP convert command, which directly caused your error.
> 
> We don't recommend running the individual tools using custom shell script because there are changes/adjustments/optimizations under-the-hood in FragPipe that most users won't know.
> 
> Please download the latest FragPipe from https://github.com/Nesvilab/FragPipe/releases/latest, and run it in GUI or command line mode.
> 
> Best,
> 
> Fengchao

thanks!

But when I use GUI, other error reported
[
[log_2023-11-10_01-46-25.txt](https://github.com/Nesvilab/FragPipe/files/13311263/log_2023-11-10_01-46-25.txt)
](url)

**Comment by fcyu** [comment]
Those are know issues and has been fixed: https://github.com/Nesvilab/FragPipe/issues/1283, https://github.com/Nesvilab/FragPipe/issues/1293, https://github.com/Nesvilab/FragPipe/issues/1275

Could you following the instruction in https://github.com/Nesvilab/FragPipe/issues/1283#issuecomment-1756054152 ?

Best,

Fengchao

**Comment by lzw150** [comment]
> Those are know issues and has been fixed: #1283, #1293, #1275
> 
> Could you following the instruction in [#1283 (comment)](https://github.com/Nesvilab/FragPipe/issues/1283#issuecomment-1756054152) ?
> 
> Best,
> 
> Fengchao

I uninstall and install the fixed EasyPQP using the following command:
pip uninstall easypqp -y
pip install git+https://github.com/Nesvilab/easypqp.git@master

Error reported:
MSBooster v1.1.11
Using 7 threads
java.lang.IllegalArgumentException: No matching mzml/mgf file for 191012_rapigest_0-1_02.pin, please check that provided mzml/mgf directories contain proper mzml/mgf files.
	at Features.PinMzmlMatcher.<init>(PinMzmlMatcher.java:93)
	at Features.MainClass.main(MainClass.java:507)
Process 'MSBooster' finished, exit code: 1
Process returned non-zero exit code, stopping

~~~~~~~~~~~~~~~~~~~~
Cancelling 207 remaining tasks

[log_2023-11-10_13-34-50.txt](https://github.com/Nesvilab/FragPipe/files/13315969/log_2023-11-10_13-34-50.txt)

**Comment by fcyu** [comment]
Updating to the latest FragPipe should be able to solve this problem.

Best,

Fengchao
